### DISCUSSION TITLE
Input

### DISCUSSION THREAD

**Discussion by bameyrick** [ONSdigital/design-system#1383] [Feedback]
Use this place to discuss the [Input Component](https://ons-design-system.netlify.com/components/input) in the [ONS Design System](https://ons-design-system.netlify.com)

**Comment by bameyrick** [comment]
Input component page needs to be updated to reference this forum issue

**Comment by bameyrick** [comment]
The `title` attribute on the input should only be applied if a prefix or suffix exists.

**Comment by bameyrick** [comment]
Inputs should be set to use Lato typeface for user input

**Comment by bameyrick** [comment]
Input width modifier doesn't work correctly when used in a `field__item`, for example the date component. Inputs will only grow as wide as their labels.

**Comment by jrbarnes9** [comment]
Use of `autocomplete` needs more work and guidance regarding its use in survey questions where the integrity of the data could suffer from pre-population

**Reply by jrbarnes9** [reply]
Raised #275

**Comment by jrbarnes9** [comment]
Inputs should have a sensible default width class (that can be overridden) for certain input types (that have a known max length e.g. postcode) so authors do not need a width each time.

**Comment by bameyrick** [comment]
Suffixed inputs should not have text right aligned as this causes issues with ZoomText.

**Comment by bameyrick** [comment]
We should hide the clear button IE11 adds to inputs as it breaks the layout of inputs with constrained widths

**Reply by jrbarnes9** [reply]
Resolved by #2374

**Comment by bameyrick** [comment]
The input component doesn't wrap itself in a field (#35) when used in mutually exclusive (#29) mode.

**Comment by bameyrick** [comment]
The suffix/prefix text for inputs is no longer vertically centred

<img width="360" alt="Screenshot 2019-07-16 at 15 02 23" src="https://user-images.githubusercontent.com/8447446/61301000-f795a780-a7da-11e9-90e0-5e2e204b2656.png">

**Comment by jrbarnes9** [comment]
UR reports users are often blind to input hint text, especially on auto-suggest lookups.  Propose to increase font-size to base.

**Comment by jrbarnes9** [comment]
Hint text shouldn't be contained within the label.  Place outside and use `aria-describedby` (see: https://design-system.service.gov.uk/components/text-input/#hint-text)

**Reply by jrbarnes9** [reply]
Resolved by https://github.com/ONSdigital/design-system/issues/2259

**Comment by jrbarnes9** [comment]
Ensure input width class numerals match number of widest character

**Reply by jrbarnes9** [reply]
Resolved by #1177

**Comment by jrbarnes9** [comment]
> Hint text shouldn't be contained within the label. Place outside and use `aria-describedby` (see: https://design-system.service.gov.uk/components/text-input/#hint-text)

Won't do this, whilst we have bounding boxes around label and hint.

**Reply by jrbarnes9** [reply]
Resolved by #2259

**Comment by jrbarnes9** [comment]
Increase mobile number width example to `input--w-9` to cater for longest case scenario of e.g. +44 08888 888 888

**Comment by jrbarnes9** [comment]
Text input focus border styling is inconsistent across devices/browsers...
On Chroms MacOS it renders as it should...
![Screenshot 2020-11-23 at 10 29 38](https://user-images.githubusercontent.com/43346934/99951957-d05bcc80-2d76-11eb-8dfe-ac8dd0d70fb5.png)

But on Chrome Windows and Chrome Android it is rendering like...
![image](https://user-images.githubusercontent.com/43346934/99952168-24ff4780-2d77-11eb-94df-8d79bbe7e613.png)

![Screenshot_20201123-102826](https://user-images.githubusercontent.com/43346934/99952063-fbdeb700-2d76-11eb-9901-5e6f6545970b.png)

**Comment by jrbarnes9** [comment]
Could we improve the way we handle number input by automatically adding separators in certain scenarios?

For example, automatically adding a comma after 4 digits are entered when asking for a number in pounds
![Screenshot 2021-09-20 at 11 29 11](https://user-images.githubusercontent.com/43346934/133988535-bc601a5c-705f-41db-90a9-bbf02d08af62.png)

**Reply by jrbarnes9** [reply]
Linked to #1947

**Comment by jrbarnes9** [comment]
Selecting an input prefix/suffix should show the tooltip and put focus on the input 

GDS have evidence that some users try to click their input prefix/suffix. ONS input prefixes/suffixes are styled similar to a secondary buttons which would further increase the affordance and the likelihood of this behaviour. Putting focus on the input when they are selected would replicate the standard label behaviour. Additionally showing the tooltip when they are selected (if is `title` applicable) would also help the discoverability of the full form label for mobile/sighted users.

![Screenshot 2022-08-18 at 15 04 28](https://user-images.githubusercontent.com/43346934/185414901-7506faaf-8d10-47d4-9e99-6559b285d1b0.png)

**Reply by jrbarnes9** [reply]
Raised #2411

**Comment by Paul-Joel** [comment]
There appears to be an issue with the interaction between accessiblePlaceholder and label>description and/or error.

accessiblePlaceholder works fine in isolation, but the text appears in the wrong location when a description is present.
![image](https://github.com/ONSdigital/design-system/assets/19411155/0f6c66bf-de90-4b3f-aea6-e25573a3fea2)

When accessiblePlaceholder is used when an error is set, the input background colour goes red instead of staying white
![image](https://github.com/ONSdigital/design-system/assets/19411155/6a08b3a1-67a5-45a1-a83f-ed8af636c3cc)

With accessiblePlaceholder, description and error set
![image](https://github.com/ONSdigital/design-system/assets/19411155/c32fa3e4-4656-40cc-900e-b14e875c99fe)

**Reply by daniellecorke** [reply]
Resolved by #2971

**Comment by AlanHONS** [comment]
I believe there's a small visual oversight with any Input that uses the `accessiblePlaceholder` option?
As the standard ONS Input placeholder text is a lighter shade of grey i.e. #707071 (see image 1) as I'd expect.
![image](https://github.com/ONSdigital/design-system/assets/114412177/f79306ce-34bc-4657-a850-0823e768f9b2)

However when you use the accessible placeholder feature, like [Search placeholder example](https://service-manual.ons.gov.uk/design-system/components/input/example-input-search-with-placeholder) the placeholder text is shown in same black #222222 as actual entered text. 

Which makes it look like text is already entered into that field (see top input vs my lower modified colour input of image 2).
![image](https://github.com/ONSdigital/design-system/assets/114412177/12d270a5-be64-4c21-8279-39347139549b)

I believe the fix is simply to add an additional declaration of `color:#707071;` to existing `.ons-label--placeholder` ?

Initially mentioned on Slack at https://onsdigital.slack.com/archives/CR7LXCA81/p1708016794325049

**Reply by daniellecorke** [reply]
#3043 raised.

**Reply by daniellecorke** [reply]
Resolved by #3043.

**Comment by AlanHONS** [comment]
Character limit checker variant on https://service-manual.ons.gov.uk/design-system/components/input#character-limit-checker doesn't actually provide a working example of this functionality. I believe this is because that example's Nunjucks configuration doesn't actually set any of mentioned charCheckLimit related attributes. Therefore Char limit functionality is not being triggered

**Reply by AlanHONS** [reply]
Discussed with Design System team. Working example already present on Filter example

**Comment by AlanHONS** [comment]
In Design System [Release 72.10.7 · ONSdigital/design-system](https://github.com/ONSdigital/design-system/releases/tag/72.10.7) you added our requested option for Textareas to display remaining allowed content in estimated words, rather than characters. 
However please confirm latest timescales to add the same functionality to Input field [Character limit checker](https://service-manual.ons.gov.uk/design-system/components/input#character-limit-checker)?
As we have a journey where we'd like to consistently represent remaining length to users in words, but have both Input and TextAreas within that same user journey

**Comment by AlanHONS** [comment]
Wanted to highlight a difference in behaviour between [Character Limit Textarea](https://service-manual.ons.gov.uk/design-system/components/textarea#character-limit) and [Character Limit Input](https://service-manual.ons.gov.uk/design-system/components/input#character-limit-checker) fields which I feel should probably be resolved.
If you specify the character limit variant of textarea then 'You have xxx characters remaining' text will display immediately and permanently under textarea field, then dynamically update it's text as the user types.
However if you do the same with an input field then the same 'You have xxx characters remaining' text will only display after the user starts to type in this field. Which results in that character count limit not being immediately visible to a user + any later page content being dynamically pushed down another line after initial page rendering.
Not currently a high priority to fix but think it would be better for [Character Limit Input](https://service-manual.ons.gov.uk/design-system/components/input#character-limit-checker) to behave consistently with same Textarea variant?

**Reply by MagdalenaLarge** [reply]
Thank you @AlanHONS  for your suggestion. 
We’ll investigate this further and let you know once we have an update.

**Reply by MagdalenaLarge** [reply]
Hello @AlanHONS 
Based on the research done by our Interaction Designer, generally we should not have a similar behaviour for both components. 

1. Text area (longer answers, higher risk)
- Best pattern: show character limit at all times
Why this is the best solution:
- Users expect to write more
- Higher chance of hitting the limit
- People often plan their response length.
- Seeing “0 / 500” sets expectations immediately
This is why your current solution for the text area makes sense.

2. Input field (short answers, low risk)
- Best pattern: show character limit on interaction (focus or first character)
Why this is the best solution:
- Most users won’t hit the limit
- Showing limits upfront adds visual noise
- It can feel restrictive before it’s necessary. 
- Users don’t need to “plan” a 20–50 character answer
This is why showing the countdown only after typing feels lighter and more appropriate.

Additionally, [Gov.uk](http://gov.uk/) do not have a character count for text input, only for text area.

I hope this explains our thought process behind the discrepancy in these components.
Thank you.
